src/transaction/atomic_replication_helper.cpp

```diff
@@ -202,46 +122,177 @@ namespace cublog
   atomic_replication_helper::is_part_of_atomic_replication (TRANID tranid) const
   {
     const auto sequence_it = m_sequences_map.find (tranid);
-    if (sequence_it == m_sequences_map.cend ())
+    if (sequence_it != m_sequences_map.cend ())
+      {
+	return true;
+      }
+
+    return false;
+  }
+
+  bool
+  atomic_replication_helper::all_log_entries_are_control (TRANID tranid) const
+  {
+    const auto sequence_it = m_sequences_map.find (tranid);
+    if (sequence_it != m_sequences_map.cend ())
       {
-	return false;
+	const atomic_log_sequence &sequence = sequence_it->second;
+	return sequence.all_log_entries_are_control ();
       }
 
+    // if no atomic sequence present, assume a positive response
     return true;
   }
 
+  LOG_LSA
+  atomic_replication_helper::get_the_lowest_start_lsa () const
+  {
+    LOG_LSA min_lsa = MAX_LSA;
+
+    for (auto const &sequence_map_iterator : m_sequences_map)
+      {
+	const atomic_log_sequence &sequence = sequence_map_iterator.second;
+	const LOG_LSA sequence_start_lsa = sequence.get_start_lsa ();
+	if (sequence_start_lsa < min_lsa)
+	  {
+	    min_lsa = sequence_start_lsa;
+	  }
+      }
+    return min_lsa;
+  }
+
   void
-  atomic_replication_helper::apply_and_unfix_atomic_replication_sequence (THREAD_ENTRY *thread_p, TRANID tranid)
+  atomic_replication_helper::append_control_log (THREAD_ENTRY *thread_p, TRANID trid,
+      LOG_RECTYPE rectype, LOG_LSA lsa, const log_rv_redo_context &redo_context)
   {
-    const auto sequence_it = m_sequences_map.find (tranid);
+    auto sequence_it = m_sequences_map.find (trid);
     if (sequence_it == m_sequences_map.end ())
       {
-	assert (false);
-	return;
+	start_sequence_internal (trid, lsa, redo_context);
+	sequence_it = m_sequences_map.find (trid);
       }
 
+    // TODO: idea, first add control log and then apply and unfix, such that apply and unfix will
+    // be able to make the decisions taking into consideration the last control log;
+    // this can be implemented later and only if needed; works as is right now
     atomic_log_sequence &sequence = sequence_it->second;
-    sequence.apply_and_unfix_sequence (thread_p);
-    m_sequences_map.erase (sequence_it);
+    sequence.apply_and_unfix (thread_p);
```

Comment:
Does this mean kinds of the following sequence are now allowed?
< start atomic sysop 1 > - < redo > - < start atomic sysop2 > - < redo > - <sysop end 1> - **< redo >** - <  sysop end 2 >.
In the above, the atomicity of sysop 1 is broken. The way I see it, there is no restriction to force the way system operations are used even if there is no case for now.

Reply:
Indeed, currently there is no distinction with regards to the actual effect of a "control" log record.
And indeed, they could even be overlapped as you describe above (even though I think this will never be the case - it will crash long before replication).

This is somewhat on purpose because in practice the following "nested" situation is encountered (notice: not the same example as the one above):
```
- < start atomic sysop 1 >
  - ..  
  - < redo VPID1 >
  - ..  
- < start atomic sysop2 >
  - ..  
  - < redo VPID1 >
  - ..  
- < sysop end 2 >
  - ..  
  - < redo >
  - ..  
- < sysop end 1 >
```
If the logic were to follow "topologically" and first apply the redos of the inner atomic sequence and then the redo's of the outer atomic sequence, the page VPID1 would be inconsistently modified (if lucky, crash, if unlucky - incorrect replication and wrong results).

Reply:
I know that all redo history is repeated in order. But, if VPID1 is unfixed after redoing (1), other threads can cut in and fix the page at (2) and (3). It seems to break the atomicity as long as I don't misunderstand.

Reply:
```
They could be overlapped as you describe above (even though I think this will never be the case - it will crash long before replication).
```
I don't understand why this sequence is impossible, and it means I may miss something..

Reply:
> I know that all REDO history has to be repeated in order. I mean if VPID1 is unfixed after redoing (1), other threads can cut in and fix the page before fixing the page for (2) and (3 - if it also modifies VPID1). It seems to break the atomicity as long as I don't misunderstand.

Yes, this is a possibility.
However:
- in the first version of the atomic replication, the pages were kept fixed by the replication from `< start atomic sysop 1 >` until `< sysop end 1 >`;
- this was deemed too costly for performance; and - based on examples seen then - there was no counter argument as to not apply log records and unfix pages at every occurrence of a "control log record"; hence the current refactoring

Reply:
I understand it is too costly, but it is one of the atomic replication's inherent attributes, the way I see it. If an atomic operation gets bigger, it will fix many pages and cost a lot. I don't get the point of why this kind of nested scenario is an exceptional case for which some records are applied and its pages are unfixed. Are they longer than others usually?

Reply:
> Are they longer than others usually?

I think they are. However I also think that we haven't yet seen the "worst of it".
For example there haven't been any tests with very large records in the schema (table records spanning multiple pages).

I see action possible in the following 3 directions:

1)
However, it is relatively simple to overload the logic to either do it as it is now (piece-wise between "control" log records) or all in one bit "proper" atomic sequence (as it was initially implemented).

2)
However, we're facing with a "chicken & egg" problem:
- currently the implementation for PTS replication is incomplete
- given that, not many tests can be added since they fail

3)
Given this, I propose to
- merge this with this "piece-wise" replication method
- wait for http://jira.cubrid.org/browse/LETS-524 to be implemented - because that will lessen the number of atomic sequence scenarios to be supported
- in the meantime I will add new logging code such that the complete atomic sequence is dumped (text representation of all "control" and "non-control" log records) at the end, before the atomic sequence is destroyed.

With this we'll be hopefully able to advance some testing scenarios that we'll allow further analysis.
We might even discover crashes caused precisely by this "piece-wise" atomic replication technique.

Reply:
OK. Let's move forward with the recognition that this is another problem that needs to be solved in the long term. We will solve it and others together. Thanks.

Reply:
> that needs to be solved in the long term

I would say "short term". :)